docs/declarative-pipelines-programming-guide.md

```diff
@@ -0,0 +1,416 @@
+---
+layout: global
+title: Spark Declarative Pipelines Programming Guide
+displayTitle: Spark Declarative Pipelines Programming Guide
+license: |
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+ 
+     http://www.apache.org/licenses/LICENSE-2.0
+ 
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+
+* Table of contents
+{:toc}
+
+## What is Spark Declarative Pipelines (SDP)?
```

Comment:
In SPIP, there was a graph to explain the high-level idea of SDP. Shall we have a graph as well in this doc?